src/cmdlib.py

```diff
@@ -50,3 +53,82 @@ def rm_allow_noent(path):
         os.unlink(path)
     except FileNotFoundError:
         pass
+
+
+def run_supermin(workdir, script=''):
+    vmpreparedir = f"{workdir}/tmp/supermin.prepare"
+    if os.path.isdir(vmpreparedir):
+        shutil.rmtree(vmpreparedir)
+    os.makedirs(vmpreparedir, exist_ok=True)
+
+    vmbuilddir = f"{workdir}/tmp/supermin.build"
+    if os.path.isdir(vmbuilddir):
+        shutil.rmtree(vmbuilddir)
+    os.makedirs(vmbuilddir, exist_ok=True)
+
+    rpms = []
+    with open(f"{libdir}/vmdeps.txt") as vmdeps:
+        for line in vmdeps:
+            if not line or line.startswith('#'):
+                continue
+            for pkg in line.split():
+                rpms.append(pkg)
+
+    run_verbose(['supermin',
+                 '--prepare',
+                 '--use-installed',
+                 '-o', f"{vmpreparedir}"] + rpms)
+
+    with open(f"{libdir}/supermin-init-prelude.sh") as prelude:
+        sm_init = prelude.read()
+
+    with open(f"{workdir}/tmp/cmd.sh", "w+") as scriptfile:
+        scriptfile.write(script)
+    initscript = f'''#!/usr/bin/env bash
+set -xeuo pipefail
+workdir={workdir}
+{sm_init}
+RC=0
+sh {workdir}/tmp/cmd.sh || RC=$?
+echo $RC > {workdir}/tmp/rc
+/sbin/fstrim -v {workdir}/cache
+/sbin/poweroff -f
+    '''
+    with open(f"{vmpreparedir}/init", "w+") as vminit:
+        vminit.write(initscript)
+    os.chmod(f"{vmpreparedir}/init", 0o755)
+    run_verbose(['tar',
+                 '-C', f"{vmpreparedir}",
+                 '-czf', f"{vmpreparedir}/init.tar.gz",
+                 '--remove-files',
+                 'init'])
+
+    run_verbose(['supermin',
+                 '--build', f"{vmpreparedir}",
+                 '--size', '5G',
+                 '-f', 'ext2',
+                 '-o', f"{vmbuilddir}"])
+
+    nproc = multiprocessing.cpu_count()
+    run_verbose(['qemu-kvm', '-nodefaults', '-nographic', '-no-reboot',
+                 '-smp', f"{nproc}",
+                 '-m', '2048',
+                 '-kernel', f"{vmbuilddir}/kernel",
+                 '-initrd', f"{vmbuilddir}/initrd",
+                 '-netdev', 'user,id=eth0,hostname=supermin',
+                 '-device', 'virtio-net-pci,netdev=eth0',
+                 '-device', 'virtio-scsi-pci,id=scsi0,bus=pci.0,addr=0x3',
+                 '-drive', f"if=none,id=drive-scsi0-0-0-0,snapshot=on,file={vmbuilddir}/root",
+                 '-device', 'scsi-hd,bus=scsi0.0,channel=0,scsi-id=0,lun=0,drive=drive-scsi0-0-0-0,id=scsi0-0-0-0,bootindex=1',
+                 '-drive', f"if=none,id=drive-scsi0-0-0-1,discard=unmap,file={workdir}/cache/cache.qcow2",
+                 '-device', 'scsi-hd,bus=scsi0.0,channel=0,scsi-id=0,lun=1,drive=drive-scsi0-0-0-1,id=scsi0-0-0-1',
+                 '-virtfs', f"local,id=workdir,path={workdir},security_model=none,mount_tag=workdir",
+                 '-serial', 'stdio', '-append', 'root=/dev/sda console=ttyS0 selinux=1 enforcing=0 autorelabel=1'])
```

Comment:
Hopefully a next step after this would be to rewrite `runcompose()/runvm()` to use this instead.